src/python/pants/core/goals/test.py

```diff
@@ -77,25 +98,33 @@ class TestResult(EngineAwareReturnType):
     __test__ = False
 
     @classmethod
-    def skip(cls, address: Address, output_setting: ShowOutput) -> TestResult:
+    def skip(
+        cls, tester_name: str, partition_description: str, output_setting: ShowOutput
+    ) -> TestResult:
         return cls(
             exit_code=None,
             stdout="",
             stderr="",
             stdout_digest=EMPTY_FILE_DIGEST,
             stderr_digest=EMPTY_FILE_DIGEST,
-            address=address,
             output_setting=output_setting,
+            tester_name=tester_name,
+            partition_description=partition_description,
             result_metadata=None,
         )
 
+    @property
+    def skipped(self) -> bool:
+        return self.exit_code is None or self.result_metadata is None
```

Comment:
Instead, it may be better to have the partition rules do the skipping. They should have all the information they need. That will make our code to run a test simpler. That is fine as a follow up improvement though.

Reply:
That was my original approach 😄 it worked for everything but the `go` backend, which uses `TestResult.skip` [here](https://github.com/pantsbuild/pants/blob/main/src/python/pants/backend/go/goals/test.py#L221-L222). I'm not familiar with how `go` works in Pants so I don't know how a `GoTestFieldSet` could end up pointing at sources that don't contain any tests, but I assume the logic is there to handle some realistic case.

Leaving the ability to mark results as skipped within test execution will (I expect) also be helpful for things like #16846 where a test tool dynamically reports that no tests were run.

Reply:
Ah, yeah Go is interesting. So Go is interesting that we don't have a test target, only `go_package`. It may or may not have tests, but we don't know until analyzing it.

We could do that analysis inside the partition rule though, and it possibly makes sense so that our batches are more accurate and don't include targets that won't actually run.

Either way, probably fine as a follow up

Reply:
🤔 I like the idea of moving the logic into the partition rule. Challenges that come to mind from looking at it:

1. We'd probably want/need to run most of the compilation/linking in the partition rule, and have the output of that be the type tracked in `GoTestRequest.SubPartition` - we'd need to generalize some of the types used in the common API to support this
2. If we're running compilation/linking within the partition rule we'd need some way of reporting failures back to the user. I'm not sure if `assert`s/`raise`s within rules are good enough or if we'd need to change the partition API to return something like a `Result` type

I'm tempted to say we figure these out as part of this PR so the API only changes once on `main`, but I'd defer to you if you think it'd be better to save for a follow-up PR

Reply:
Yeah, I think it's better as we have it now actually. Thanks for thinking that through!